File: tests/test_m6_gguf_dequant.c
#include "q38_gguf.h"
#include "q38_quant.h"

#include <inttypes.h>
#include <stdint.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct {
    const char *name;
    uint64_t row;
} fixture_spec;

static const fixture_spec fixtures[] = {
    {"model.language_model.embed_tokens.weight", 9419},
    {"model.language_model.layers.1.ple.key_proj.weight", 0},
    {"model.language_model.layers.2.mlp.experts.gate_up_proj", 75 * 1280},
};

static uint64_t fnv1a(const unsigned char *bytes, size_t count) {
    uint64_t hash = 1469598103934665603ULL;
    for (size_t i = 0; i < count; ++i) {
        hash ^= bytes[i];
        hash *= 1099511628211ULL;
    }
    return hash;
}

static float bf16(uint16_t bits) {
    uint32_t raw = (uint32_t)bits << 16;
    float value;
    memcpy(&value, &raw, sizeof(value));
    return value;
}

static bool decode_row(const q38_gguf *model, const q38_tensor *tensor,
                      uint64_t row, float *values, size_t count) {
    uint64_t rows = 1;
    for (uint32_t i = 0; i + 1 < tensor->ndim; ++i) rows *= tensor->dim[i];
    const size_t cols = (size_t)tensor->dim[tensor->ndim - 1];
    if (row >= rows || count != cols || tensor->bytes % rows != 0) return false;
    const size_t row_bytes = (size_t)(tensor->bytes / rows);
    const unsigned char *raw = q38_gguf_tensor_data(model, tensor);
    if (!raw) return false;
    raw += row * row_bytes;
    if (tensor->type == 30) {
        for (size_t i = 0; i < cols; ++i) {
            uint16_t bits;
            memcpy(&bits, raw + 2 * i, sizeof(bits));
            values[i] = bf16(bits);
        }
        return true;
    }
    if (tensor->type == 8) {
        if (cols % 32 != 0 || row_bytes != (cols / 32) * 34) return false;
        for (size_t i = 0; i < cols; ++i) {
            const unsigned char *block = raw + (i / 32) * 34;
            uint16_t bits;
            memcpy(&bits, block, sizeof(bits));
            values[i] = q38_half_to_float(bits) *
                (float)((const int8_t *)(block + 2))[i % 32];
        }
        return true;
    }
    if (tensor->type == 10) {
        if (cols % Q38_QUANT_QK_K != 0) return false;
        return q38_quant_dequantize_row(
            tensor->type, raw, cols / Q38_QUANT_QK_K, values, cols, NULL, 0);
    }
    return false;
}

int main(int argc, char **argv) {
    if (argc != 3) {
        fprintf(stderr, "usage: %s model.gguf output.json\n", argv[0]);
        return 2;
    }
    char error[256];
    q38_gguf *model = q38_gguf_open(argv[1], error, sizeof(error));
    if (!model) {
        fprintf(stderr, "open: %s\n", error);
        return 1;
    }
    FILE *out = fopen(argv[2], "w");
    if (!out) {
        perror("output");
        q38_gguf_close(model);
        return 1;
    }
    fputs("{\"format\":\"q38-m6-gguf-dequant-fixtures-v1\",\"fixtures\":[",
          out);
    for (size_t f = 0; f < sizeof(fixtures) / sizeof(fixtures[0]); ++f) {
        const fixture_spec *spec = &fixtures[f];
        const q38_tensor *tensor = NULL;
        for (uint64_t i = 0; i < model->n_tensors; ++i) {
            q38_tensor *candidate = &model->tensors[i];
            if (strlen(spec->name) == candidate->name.len &&
                memcmp(spec->name, candidate->name.ptr,
                       candidate->name.len) == 0) {
                tensor = candidate;
                break;
            }
        }
        if (!tensor || tensor->ndim == 0) {
            fprintf(stderr, "missing fixture tensor: %s\n", spec->name);
            fclose(out);
            q38_gguf_close(model);
            return 1;
        }
        const size_t cols = (size_t)tensor->dim[tensor->ndim - 1];
        uint64_t rows = 1;
        for (uint32_t i = 0; i + 1 < tensor->ndim; ++i) rows *= tensor->dim[i];
        const size_t row_bytes = (size_t)(tensor->bytes / rows);
        float *values = calloc(cols, sizeof(*values));
        const unsigned char *payload = q38_gguf_tensor_data(model, tensor);
        if (!values || !payload || !decode_row(model, tensor, spec->row,
                                                values, cols)) {
            fprintf(stderr, "cannot decode fixture tensor: %s\n", spec->name);
            free(values);
            fclose(out);
            q38_gguf_close(model);
            return 1;
        }
        if (f) fputc(',', out);
        fprintf(out, "{\"tensor\":\"%s\",\"type\":%u,\"ndim\":%u,\"shape\":[",
                spec->name, tensor->type, tensor->ndim);
        for (uint32_t i = 0; i < tensor->ndim; ++i) {
            if (i) fputc(',', out);
            fprintf(out, "%" PRIu64, tensor->dim[i]);
        }
        fprintf(out, "],\"row\":%" PRIu64 ",\"row_bytes\":%zu,"
                    "\"raw_fnv1a\":\"%016" PRIx64 "\",\"values\":[",
                spec->row, row_bytes, fnv1a(payload + spec->row * row_bytes,
                                            row_bytes));
        for (size_t i = 0; i < cols; ++i) {
            if (i) fputc(',', out);
            fprintf(out, "%.9g", values[i]);
        }
        fputs("]}", out);
        free(values);
    }
    fputs("]}\n", out);
    fclose(out);
    q38_gguf_close(model);
    return 0;
}

File: tests/test_m6_gguf_dequant.py
#!/usr/bin/env python3
"""Fail-closed byte-level checks for the GGUF dequantization fixtures."""

from __future__ import annotations

import hashlib
import json
import math
import struct
import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parents[1]))
from tools.m6_quant_matched_reference import GGUF


EXPECTED = {
    "model.language_model.embed_tokens.weight": (30, (248320, 2560), 9419),
    "model.language_model.layers.1.ple.key_proj.weight": (8, (10240, 2560), 0),
    "model.language_model.layers.2.mlp.experts.gate_up_proj": (
        10, (512, 1280, 2560), 75 * 1280),
}


def bf16(bits: int) -> float:
    return struct.unpack("<f", struct.pack("<I", bits << 16))[0]


def q2_oracle(raw: memoryview, cols: int) -> list[float]:
    result = [0.0] * cols
    for block_index in range(cols // 256):
        base = block_index * 84
        scales = raw[base:base + 16]
        q = raw[base + 16:base + 80]
        d = struct.unpack_from("<e", raw, base + 80)[0]
        dmin = struct.unpack_from("<e", raw, base + 82)[0]
        at = block_index * 256
        scale_index = 0
        for half in (0, 128):
            qbase = half // 4
            for j in range(4):
                shift = 2 * j
                for part, offset in enumerate((0, 16)):
                    scale = scales[scale_index]
                    scale_index += 1
                    dl = d * (scale & 15)
                    ml = dmin * (scale >> 4)
                    start = at + half + j * 32 + part * 16
                    for l in range(16):
                        result[start + l] = (
                            dl * ((q[qbase + offset + l] >> shift) & 3) - ml
                        )
    return result


def oracle(reader: GGUF, name: str, row: int) -> tuple[bytes, list[float]]:
    descriptor = reader.descriptor(name)
    shape = descriptor["shape"]
    rows, cols = math.prod(shape[:-1]), shape[-1]
    block, block_bytes = {8: (32, 34), 10: (256, 84)}.get(
        descriptor["type"], (1, 2)
    )
    row_bytes = cols // block * block_bytes
    start = descriptor["offset"] + row * row_bytes
    raw = memoryview(reader.mm)[start:start + row_bytes]
    if descriptor["type"] == 30:
        result = [
            bf16(struct.unpack_from("<H", raw, i * 2)[0])
            for i in range(cols)
        ]
    elif descriptor["type"] == 8:
        result = [
            struct.unpack_from("<e", raw, (i // 32) * 34)[0] *
            struct.unpack_from("<b", raw, (i // 32) * 34 + 2 + i % 32)[0]
            for i in range(cols)
        ]
    elif descriptor["type"] == 10:
        result = q2_oracle(raw, cols)
    else:
        raise AssertionError(f"unsupported fixture type: {descriptor['type']}")
    return bytes(raw), result


def main() -> None:
    if len(sys.argv) != 3:
        raise SystemExit("usage: test_m6_gguf_dequant.py model.gguf fixture.json")
    model, fixture_path = Path(sys.argv[1]), Path(sys.argv[2])
    report = json.loads(fixture_path.read_text())
    if report.get("format") != "q38-m6-gguf-dequant-fixtures-v1":
        raise SystemExit("fixture format mismatch")
    reader = GGUF(model)
    try:
        fixtures = report.get("fixtures")
        if not isinstance(fixtures, list) or len(fixtures) != len(EXPECTED):
            raise SystemExit("fixture set is incomplete")
        for fixture in fixtures:
            name = fixture.get("tensor")
            if name not in EXPECTED:
                raise SystemExit(f"unexpected fixture tensor: {name}")
            expected_type, expected_shape, expected_row = EXPECTED[name]
            descriptor = reader.descriptor(name)
            if (fixture.get("type"), tuple(fixture.get("shape", [])),
                    fixture.get("row")) != (
                        expected_type, expected_shape, expected_row):
                raise SystemExit(f"{name}: type/shape/row mismatch")
            raw, expected = oracle(reader, name, expected_row)
            expected_stride = len(raw)
            if fixture.get("row_bytes") != expected_stride:
                raise SystemExit(f"{name}: row stride mismatch")
            if fixture.get("raw_sha256") is None:
                fixture["raw_sha256"] = hashlib.sha256(raw).hexdigest()
            elif fixture["raw_sha256"] != hashlib.sha256(raw).hexdigest():
                raise SystemExit(f"{name}: raw checksum mismatch")
            actual = fixture.get("values")
            if not isinstance(actual, list) or len(actual) != len(expected):
                raise SystemExit(f"{name}: decoded row length mismatch")
            if any(not math.isclose(float(a), b, rel_tol=1e-6, abs_tol=1e-6)
                   for a, b in zip(actual, expected)):
                raise SystemExit(f"{name}: q38/oracle decoded values diverge")
            if int(fixture.get("raw_fnv1a", "0"), 16) == 0:
                raise SystemExit(f"{name}: missing q38 raw checksum")
        fixture_path.write_text(json.dumps(report, indent=2) + "\n")
    finally:
        reader.close()
    print("test_m6_gguf_dequant: BF16/Q8_0/Q2_K GGUF fixtures passed")


if __name__ == "__main__":
    main()
